internal/dcache/rpc/client/functions.go

```diff
@@ -465,24 +534,48 @@ func PutChunkDC(ctx context.Context, targetNodeID string, req *models.PutChunkDC
 					log.Err("rpc_client::PutChunkDC: resetAllRPCClients failed for node %s: %v",
 						targetNodeID, err1)
 					//
-					// Connection refused and timeout are the only viable errors.
+					// Connection refused and timeout are the only viable errors, or resetAllRPCClients()
+					// may refuse to create new connections if the node is marked negative.
 					// Assert to know if anything else happens.
 					//
-					common.Assert(rpc.IsConnectionRefused(err1) || rpc.IsTimedOut(err1), err1)
+					common.Assert(rpc.IsConnectionRefused(err1) ||
+						rpc.IsTimedOut(err1) ||
+						errors.Is(err1, NegativeNodeError), err1)
 					return nil, err
 				}
 
 				// Retry PutChunkDC once more with fresh connection.
 				continue
 			} else if rpc.IsTimedOut(err) {
 				//
-				// If the error is due to a timeout (either node is down or cannot be reached over the n/w),
-				// we delete this RPC client and all existing clients in the pool as reusing it will most
-				// likely fail with the same error.
+				// If we get timeout error in PutChunkDC(), it means that one/more of the downstream
+				// nodes/connections are down/bad. We cannot say for sure which node is down or which
+				// connection is bad.
+				// So, we mark all the RVs (next-hop as well as next RVs in chain) as iffy RVs.
+				// This prevents other threads from also timing out if they are calling PutChunkDC to same RVs.
 				//
-				err1 := cp.deleteAllRPCClients(client)
+				cp.addIffyRvId(req.Request.Chunk.Address.RvID)
+
+				// Add the next RVs to the iffy RVs map.
+				for _, nextRV := range req.NextRVs {
+					cp.addIffyRvName(nextRV)
+				}
+
+				//
+				// In PutChunkDC we can get timeout because of bad connection between the downstream nodes,
+				// and not necessarily between the client node and target node. In this case, we retry WriteMV using
+				// the OriginatorSendsToAll strategy. So, to be safe we don't delete the connections if we get
+				// timeout error in PutChunkDC. In the PutChunk using OriginatorSendsToAll strategy fails using
+				// timeout, we then delete the connections.
+				//
+				// We do resetAllRPCClients() here because if the connection between client and target node is good,
+				// the response from target node will eventually return after the timeout error. In this case, we
+				// cannot reuse the same which was timed out as it will result in ambiguous behavior.
```

Comment:
isn't it better to reset just this client since the error is received on this connection.
Others mostly will be fine or max one of the other connections would need to be reset but that we can do when we know for sure with OriginatorSendsToAll. Anyways, if it's bad the reset itself will timeout, so we don't save on time either.